File: shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.py
class Solution:
    def shortestPathBinaryMatrix(self, grid: List[List[int]]) -> int:
        if grid[0][0] == 1:
            return -1
        elif grid[-1][-1] == 1:
            return -1
        if len(grid) == len(grid[0]) and len(grid) == 1:
            return 1
        directions = [[0,1],[1,0],[-1,0],[0,-1], [1,1], [-1,-1], [1,-1], [-1, 1]]
        grid[0][0] = 1
        queue = [(0,0,1)]
        while queue:
            i, j, depth = queue.pop(0)
            if (i,j) == (len(grid)-1, len(grid[0])-1):
                return depth
            for d in directions:
                ix, jx, depthx = i+d[0], j+d[1], depth+1
                if ix>=0 and jx>=0 and ix<len(grid) and jx<len(grid[0]):
                    if grid[ix][jx] == 0:
                        grid[ix][jx] = depthx
                        queue.append((ix,jx,depthx))
                        
        return -1

File: shortest-path-in-binary-matrix/shortest-path-in-binary-matrix.cpp
class Solution {
public:
    int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
        if(grid[0][0] == 1 || grid[grid.size()-1][grid[0].size()-1] == 1) return -1;
        
        int directions[][2] = {{0,1}, {1,0}, {0,-1}, {-1,0}, {1,1}, {-1,-1}, {-1,1}, {1,-1}};
        grid[0][0] = 1;
            
        queue<array<int, 2>> q;
        q.push({0, 0});
        while(q.size() > 0) {
            auto [i, j] = q.front(); q.pop();
            if(i == grid.size()-1 && j == grid[0].size()-1) {
                return grid[i][j];
            }
            for(auto &d: directions) {
                int ix = i+d[0], jx = j+d[1];
                if (ix>=0 && jx>=0 && ix<grid.size() && jx<grid[0].size()) {
                    if(grid[ix][jx] == 0) {
                        grid[ix][jx] = grid[i][j] + 1;
                        q.push({ix, jx});
                    }
                }
            }
        }
        return -1;
    }
};